src/data/cel.rs

```diff
@@ -0,0 +1,499 @@
+use crate::data::get_attribute;
+use crate::data::property::Path;
+use cel_interpreter::objects::{Map, ValueType};
+use cel_interpreter::{Context, Value};
+use cel_parser::{parse, Expression as CelExpression, Member, ParseError};
+use chrono::{DateTime, FixedOffset};
+use serde_json::Value as JsonValue;
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+use std::sync::OnceLock;
+
+#[derive(Clone, Debug)]
+pub struct Expression {
+    attributes: Vec<Attribute>,
+    expression: CelExpression,
+}
+
+impl Expression {
+    pub fn new(expression: &str) -> Result<Self, ParseError> {
+        let expression = parse(expression)?;
+
+        let mut props = Vec::with_capacity(5);
+        properties(&expression, &mut props, &mut Vec::default());
+
+        let attributes = props
+            .into_iter()
+            .map(|tokens| {
+                let path = Path::new(tokens);
+                known_attribute_for(&path).unwrap_or(Attribute {
+                    path,
+                    cel_type: None,
+                })
+            })
+            .collect();
+
+        Ok(Self {
+            attributes,
+            expression,
+        })
+    }
+
+    pub fn eval(&self) -> Value {
+        let mut ctx = Context::default();
+        let Map { map } = self.build_data_map();
+
+        // if expression was "auth.identity.anonymous",
+        // {
+        //   "auth": { "identity": { "anonymous": true } }
+        // }
+        for binding in ["request", "metadata", "source", "destination", "auth"] {
+            ctx.add_variable_from_value(
+                binding,
+                map.get(&binding.into()).cloned().unwrap_or(Value::Null),
+            );
+        }
+        Value::resolve(&self.expression, &ctx).expect("Cel expression couldn't be evaluated")
+    }
+
+    fn build_data_map(&self) -> Map {
+        data::AttributeMap::new(self.attributes.clone()).into()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Predicate {
+    expression: Expression,
+}
+
+impl Predicate {
+    pub fn new(predicate: &str) -> Result<Self, ParseError> {
+        Ok(Self {
+            expression: Expression::new(predicate)?,
+        })
+    }
+
+    pub fn test(&self) -> bool {
+        match self.expression.eval() {
+            Value::Bool(result) => result,
+            _ => false,
+        }
+    }
+}
+
+pub struct Attribute {
```

Comment:
Well yes, but decided _not_ because it has the concept of _known attribute types_ that are CEL. So it could actually be `CelTypedAttribute` instead...

Reply:
Which btw is why it's in the `cel` ~file~ module.